objectivec/GPBMessage_PackagePrivate.h

```diff
@@ -42,6 +42,103 @@
+    ".long 5 *" GPB_ASM_PTRSIZE "\n"                               \
+    ".long 5 *" GPB_ASM_PTRSIZE "\n"                               \
+    GTP_ASM_ONLY_LP64(".long 0 \n")                                \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "a" #name " \n"                                    \
+    GPB_ASM_PTR "__OBJC_$_CLASS_METHODS_" #name " \n"              \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "0 \n"                                             \
+                                                                   \
+    ".section __DATA,__objc_const, regular \n"                     \
+    ".p2align" GTP_ALIGN_SIZE "\n"                                 \
+    "__OBJC_CLASS_RO_$_" #name ": \n"                              \
+    ".long 0x80 \n"                                                \
+    ".long" GPB_ASM_PTRSIZE "\n"                                   \
+    ".long" GPB_ASM_PTRSIZE "\n"                                   \
+    GTP_ASM_ONLY_LP64(".long 0 \n")                                \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "a" #name " \n"                                    \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "0 \n"                                             \
+    GPB_ASM_PTR "0 \n"                                             \
+                                                                   \
+    ".section __DATA, __objc_classlist, regular, no_dead_strip \n" \
+    ".p2align" GTP_ALIGN_SIZE "\n"                                 \
+    GPB_ASM_PTR "_OBJC_CLASS_$_" #name " \n"                       \
+                                                                   \
+    ".section __TEXT, __objc_classname, cstring_literals \n"       \
+    "a" #name ": .asciz \"" #name "\" \n"                          \
+                                                                   \
+    ".ifndef aDescriptor \n"                                       \
+    ".section __TEXT, __objc_methname, cstring_literals \n"        \
+    "aDescriptor: .asciz \"descriptor\" \n"                        \
+    ".endif \n"                                                    \
+                                                                   \
+    ".ifndef a1608 \n"                                             \
+    ".section __TEXT, __objc_methtype, cstring_literals \n"        \
+    "a1608: .asciz \"@16@0:8\" \n"                                 \
+    ".endif \n"                                                    \
+  )
+
```

Comment:
Should/Can we undef the CPP macros that are impl details inside the main one?  Should we name them in any way to make sure folks don't use them?

Reply:
We can't undef them because they need to be defined when the macro expands in-place. I figured that the GPB names (and hiding them in Private) was sufficient deterrent.